File: backend/djangobackend/djangobackend/permissions.py
from rest_framework import permissions
from apps.users.models import UserRoles


class IsStaffLevel(permissions.BasePermission):
    """
    Permission for Admin or Gerente users only.
    These users have full access to all resources.
    """
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        return request.user.rol in [UserRoles.ADMIN, UserRoles.MANAGER] or request.user.is_superuser

    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        return request.user.rol in [UserRoles.ADMIN, UserRoles.MANAGER] or request.user.is_superuser


class IsVendedor(permissions.BasePermission):
    """
    Permission for Vendedor users - READ only access.
    Vendedor can only view (GET, HEAD, OPTIONS).
    """
    def has_permission(self, request, view):
        if not request.user.is_authenticated:
            return False
        
        # Vendedor can only read
        if request.user.rol == UserRoles.SALES:
            return request.method in permissions.SAFE_METHODS
        
        return False

    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        
        # Vendedor can only read
        if request.user.rol == UserRoles.SALES:
            return request.method in permissions.SAFE_METHODS
        
        return False


class IsStaffOrVendedorReadOnly(permissions.BasePermission):
    """
    Composite permission:
    - Admin and Gerente have full access (via IsStaffLevel)
    - Vendedor has read-only access (via IsVendedor)
    
    Used for: Materias Primas, Products
    """
    def has_permission(self, request, view):
        # Check if user passes either permission
        staff_permission = IsStaffLevel()
        vendedor_permission = IsVendedor()
        
        return (staff_permission.has_permission(request, view) or 
                vendedor_permission.has_permission(request, view))

    def has_object_permission(self, request, view, obj):
        staff_permission = IsStaffLevel()
        vendedor_permission = IsVendedor()
        
        return (staff_permission.has_object_permission(request, view, obj) or 
                vendedor_permission.has_object_permission(request, view, obj))

class IsStafforVendedorReadandCreate(permissions.BasePermission):
    """
    Composite permission:
    - Admin and Gerente have full access (via IsStaffLevel)
    - Vendedor has read and create access (via IsVendedor)
    
    Used for: Materias Primas, Products
    """

    def has_permission(self, request, view):
        # 1. Check for Staff Level (Admin/Manager) - Full Access
        if IsStaffLevel().has_permission(request, view):
            return True
            
        # 2. Check for Vendedor - Read & Create Access Only
        if request.user.is_authenticated and request.user.rol == UserRoles.SALES:
            # Allow SAFE methods (GET, HEAD, OPTIONS) OR POST (Create)
            return request.method in permissions.SAFE_METHODS or request.method == 'POST'
            
        return False

    def has_object_permission(self, request, view, obj):
        # 1. Check for Staff Level - Full Access
        if IsStaffLevel().has_object_permission(request, view, obj):
             return True

        # 2. Check for Vendedor - Read Only object access
        # (They can create via POST in has_permission, but usually can't EDIT/DELETE objects)
        if request.user.is_authenticated and request.user.rol == UserRoles.SALES:
            return request.method in permissions.SAFE_METHODS
            
        return False

class IsStaffLevelOnly(permissions.BasePermission):
    """
    Permission that DENIES access to Vendedor users completely.
    Only Admin and Gerente have access.
    
    Used for: Recetas, Production, Purchases, Reports, Transformation
    """
    def has_permission(self, request, view):
        staff_permission = IsStaffLevel()
        return staff_permission.has_permission(request, view)

    def has_object_permission(self, request, view, obj):
        staff_permission = IsStaffLevel()
        return staff_permission.has_object_permission(request, view, obj)


class IsAllUsersCRUD(permissions.BasePermission):
    """
    Permission that allows all authenticated users (Admin, Gerente, Vendedor) 
    to perform CRUD operations.
    
    Used for: Clients, Sales Orders, Dashboard, Notifications, Point of Sale
    """
    def has_permission(self, request, view):
        return request.user.is_authenticated

    def has_object_permission(self, request, view, obj):
        return request.user.is_authenticated

File: frontend/panaderia-interfaz/src/features/Authentication/lib/permissions.ts
import type { Permissions } from "@/features/Authentication/types/types";

export const PERMISSIONS: Permissions = {
    // Vendedor: READ-only for materias primas and products, CRUD for clients/sales/notifications
    vendedor: [
        // View only - Materias Primas & Products
        'view:materias_primas',
        'view:productos_elaborados',
        'view:productos_reventa',

        // Full CRUD - Clients
        'view:clientes',
        'add:clientes',
        'edit:clientes',
        'delete:clientes',

        // Full CRUD - Sales Orders
        'view:ordenes_venta',
        'add:ordenes_venta',
        'edit:ordenes_venta',
        'delete:ordenes_venta',

        // Full CRUD - Notifications
        'view:notificaciones',
        'add:notificaciones',
        'edit:notificaciones',
        'delete:notificaciones',

        // View - Dashboard
        'view:dashboard',

        // Future: Point of Sale
        'view:punto_venta',
        
        // Lots
        'view:lots'
    ],

    // Gerente: Full access to everything (same as Admin)
    gerente: [
        // Materias Primas
        'view:materias_primas',
        'add:materias_primas',
        'edit:materias_primas',
        'delete:materias_primas',

        // Productos Elaborados
        'view:productos_elaborados',
        'add:productos_elaborados',
        'edit:productos_elaborados',
        'delete:productos_elaborados',

        // Productos Reventa
        'view:productos_reventa',
        'add:productos_reventa',
        'edit:productos_reventa',
        'delete:productos_reventa',

        // Clientes
        'view:clientes',
        'add:clientes',
        'edit:clientes',
        'delete:clientes',

        // Ordenes de Venta
        'view:ordenes_venta',
        'add:ordenes_venta',
        'edit:ordenes_venta',
        'delete:ordenes_venta',

        // Recetas
        'view:recetas',
        'add:recetas',
        'edit:recetas',
        'delete:recetas',

        // Producción
        'view:produccion',
        'add:produccion',
        'edit:produccion',
        'delete:produccion',

        // Compras
        'view:compras',
        'add:compras',
        'edit:compras',
        'delete:compras',

        // Ordenes de Compra
        'view:ordenes_compra',
        'add:ordenes_compra',
        'edit:ordenes_compra',
        'delete:ordenes_compra',

        // Proveedores
        'view:proveedores',
        'add:proveedores',
        'edit:proveedores',
        'delete:proveedores',

        // Transformación
        'view:transformacion',
        'add:transformacion',
        'edit:transformacion',
        'delete:transformacion',

        // Notificaciones
        'view:notificaciones',
        'add:notificaciones',
        'edit:notificaciones',
        'delete:notificaciones',

        // Dashboard
        'view:dashboard',

        // Reportes
        'view:reportes',
        'add:reportes',
        'edit:reportes',
        'delete:reportes',

        // Punto de Venta
        'view:punto_venta',
        'add:punto_venta',
        'edit:punto_venta',
        'delete:punto_venta',

        // Lots
        'view:lots',
        'add:lots',
        'edit:lots',
        'delete:lots',
    ],

    // Admin: Full access to everything (same as Gerente)
    admin: [
        // Materias Primas
        'view:materias_primas',
        'add:materias_primas',
        'edit:materias_primas',
        'delete:materias_primas',

        // Productos Elaborados
        'view:productos_elaborados',
        'add:productos_elaborados',
        'edit:productos_elaborados',
        'delete:productos_elaborados',

        // Productos Reventa
        'view:productos_reventa',
        'add:productos_reventa',
        'edit:productos_reventa',
        'delete:productos_reventa',

        // Clientes
        'view:clientes',
        'add:clientes',
        'edit:clientes',
        'delete:clientes',

        // Ordenes de Venta
        'view:ordenes_venta',
        'add:ordenes_venta',
        'edit:ordenes_venta',
        'delete:ordenes_venta',

        // Recetas
        'view:recetas',
        'add:recetas',
        'edit:recetas',
        'delete:recetas',

        // Producción
        'view:produccion',
        'add:produccion',
        'edit:produccion',
        'delete:produccion',

        // Compras
        'view:compras',
        'add:compras',
        'edit:compras',
        'delete:compras',

        // Ordenes de Compra
        'view:ordenes_compra',
        'add:ordenes_compra',
        'edit:ordenes_compra',
        'delete:ordenes_compra',

        // Proveedores
        'view:proveedores',
        'add:proveedores',
        'edit:proveedores',
        'delete:proveedores',

        // Transformación
        'view:transformacion',
        'add:transformacion',
        'edit:transformacion',
        'delete:transformacion',

        // Notificaciones
        'view:notificaciones',
        'add:notificaciones',
        'edit:notificaciones',
        'delete:notificaciones',

        // Dashboard
        'view:dashboard',

        // Reportes
        'view:reportes',
        'add:reportes',
        'edit:reportes',
        'delete:reportes',

        // Punto de Venta
        'view:punto_venta',
        'add:punto_venta',
        'edit:punto_venta',
        'delete:punto_venta',

        // Lots
        'view:lots',
        'add:lots',
        'edit:lots',
        'delete:lots',

        // User Management (Admin only)
        'view:usuarios',
        'add:usuarios',
        'edit:usuarios',
        'delete:usuarios',
    ],
};